packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+    }
+
+    return this.getElementDisplayStatus({
+      elementId: node.key.instanceKeys[0].id,
+      modelId,
+      categoryId,
+      hasChildren: node.children,
+    });
+  }
+
+  private getSubjectNodeVisibilityStatus(node: HierarchyNode, subjectIds: Observable<Id64String>): Observable<VisibilityStatus> {
+    const result = defer(() => {
+      if (!this._props.viewport.view.isSpatialView()) {
+        return of(createVisibilityStatus("disabled", "subject.nonSpatialView"));
+      }
+
+      const statuses = subjectIds.pipe(
+        mergeMap((subjectId) => this._queryHandler.querySubjectModels(subjectId)),
+        mergeMap((modelId) => this.getModelVisibilityStatus(modelId)),
+      );
+
+      // const provider = this.filteredDataProvider;
+      // const statuses =
+      //   provider && !provider.nodeMatchesFilter(node)
+      //     ? from(provider.getNodes(node)).pipe(
+      //         concatAll(),
+      //         mergeMap((filteredNode) => this.getVisibilityStatusObs(filteredNode)),
+      //       )
+      //     : subjectIds.pipe(
+      //         mergeMap((subjectId) => this._queryHandler.querySubjectModels(subjectId)),
+      //         mergeMap((modelId) => this.getModelVisibilityStatus(modelId)),
+      //       );
+
+      return statuses.pipe(
+        map((x) => x.state),
+        getVisibilityStatusFromTreeNodeChildren({
+          visible: "subject.allModelsVisible",
+          hidden: "subject.allModelsHidden",
+          partial: "subject.someModelsHidden",
+        }),
+      );
+    });
+
+    const ovr = this._props.overrides?.getSubjectNodeVisibility;
+    return ovr
+      ? subjectIds.pipe(
+          toSet(),
+          mergeMap(async (ids) => ovr(this.createOverrideProps({ node, ids }, result))),
+        )
+      : result;
+  }
+
+  private getModelVisibilityStatus(modelId: Id64String): Observable<VisibilityStatus> {
+    const result = defer(() => this.getModelVisibilityStatusImpl(modelId));
+    const ovr = this._props.overrides?.getModelDisplayStatus;
+    return ovr ? from(ovr(this.createOverrideProps({ id: modelId }, result))) : result;
+  }
+
+  private getModelVisibilityStatusImpl(modelId: Id64String): Observable<VisibilityStatus> {
+    const viewport = this._props.viewport;
+    if (!viewport.view.isSpatialView()) {
+      return createVisibilityStatusObs("disabled", "model.nonSpatialView");
+    }
+
+    if (!viewport.view.viewsModel(modelId)) {
+      return createVisibilityStatusObs("hidden", "model.hiddenThroughModelSelector");
+    }
+
+    return this._queryHandler.queryModelCategories(modelId).pipe(
+      map((categoryId) => this.getDefaultCategoryVisibilityStatus(categoryId, modelId)),
+      map((x) => x.state),
+      getVisibilityFromTreeNodeChildren,
+      mergeMap((visibilityByCategories) => {
+        // istanbul ignore if
+        if (visibilityByCategories === "empty") {
+          return createVisibilityStatusObs("visible");
+        }
```

Comment:
I think it can be possible. You previously said that we allow showing models with no elements. If model has no elements that it has no categories, does it?

Reply:
Ah, right. In that case we need a test rather than `istanbul ignore` :)

Reply:
Added a test. Also enabled coverage for `ModelsTreeVisibilityHandler` and `AlwaysAndNeverDrawnElementInfo`